added PhoneNumberType check for twig

toooni opened the pull request:
solves #30 

there is a BC break when using PHP template because the helper is renamed to `phone_number_helper` from `phone_number_format`.

JHGitty commented:
What's about merging this?

thewilkybarkid commented:
Looks ok, but we can't have a BC break. It should be a separate extension.

toooni commented:
@thewilkybarkid why not? I don't think it's good to create an extension for every single function. Maybe I can create an alias `phone_number_format`? Then then we can avoid a BC break..

thewilkybarkid commented:
This is quite a different operation. I don't particularly like the `PhoneNumberUtil` for the same reason (it handles parsing, formatting etc).

Anyhow, that's not too important. The BC break needs to be fixed by either aliasing it or having separate extensions.

toooni commented:
@thewilkybarkid i fixed the BC-break by adding an alias and rebased the whole PR.

The handling of `PhoneNumberUtil` isn't perfect, yes. But I do think that especially the usage of one single twig extension is the right thing to do. The whole point of a Twig-Extension is that you can add multiple functions, methods or globals.

JHGitty commented:
I think you should not add too much Twig extensions because currently there are no lazy loading features  for Twig (and as I know there is no planned feature for that because it is not easy possible?). It is a great idea to put a lot of functions and filters to one single Twig extension.

toooni commented:
@thewilkybarkid i have rebased the PR. I hope you reconsider merging it. (The travis-ci error does not seem to be the origin of this PR)

toooni commented:
@thewilkybarkid rebased the PR once again to have travis-ci checks passed as well.

toooni commented:
@thewilkybarkid what can I do?

thewilkybarkid commented on Resources/config/services.xml:
```diff
@@ -29,37 +28,29 @@
+            <tag name="jms_serializer.handler" type="libphonenumber\PhoneNumber" direction="deserialization" format="json" method="deserializePhoneNumberFromJson"/>
+            <tag name="jms_serializer.handler" type="libphonenumber\PhoneNumber" direction="serialization" format="xml" method="serializePhoneNumber"/>
+            <tag name="jms_serializer.handler" type="libphonenumber\PhoneNumber" direction="deserialization" format="xml" method="deserializePhoneNumberFromXml"/>
+            <tag name="jms_serializer.handler" type="libphonenumber\PhoneNumber" direction="serialization" format="yml" method="serializePhoneNumber"/>
```
Revert these, nothing has changed and it makes the diff hard to read. (There are lots of other non-changes that need reverting too, eg new line removal.)

thewilkybarkid commented:
Renaming classes, container parameters etc are all BC breaks.

It should be possible to leave the existing names and mark them as deprecated.

toooni commented:
@thewilkybarkid 
- readded and deprecated the classes
- readded the container parameters (already deprecated, not used anywhere)
- resetted the coding style changes

toooni commented:
@thewilkybarkid should i resolve the conflicts or isn't there a chance this is going to get merged?

toooni commented:
@rh389 I saw that you merged the last PR. Can you help here?

rh389 commented:
@toooni I'm reluctant to maintain this lib because I don't even work with PHP any more, but since you asked...

It looks like you've removed the BC breaks, which is great, but I'd be happier if there were a couple of simple tests to confirm that:

```php
public function testDeprecatedClassName() {
  ...
  $extension = new PhoneNumberFormatExtension(...)
  $this->assertInstanceOf('...PhoneNumberHelperExtension', $extension)
}
```
and
```php
  // Assert deprecated 'phone_number_format' alias
  $this->assertServiceHasTag(
    'misd_phone_number.templating.helper',
    'templating.helper',
    array('alias' => 'phone_number_format')
  );
}
```

If you can add those and resolve conflicts, and if @thewilkybarkid doesn't object before the end of this week, I'll merge.

toooni commented:
@rh389 thank you for your response!
I rebased the branch and added the tests as requested. Since the the twig strategy had changed I also changed the new `phone_number_is_type` from function to filter. But I took the liberty to also leave the function and left it as deprecated.

rh389 commented:
Thanks. Looks good to me - I'll merge on Friday if @thewilkybarkid doesn't object before then. (Ping me if I don't).

toooni commented:
@rh389 thank you.. 
btw. the failing tests are just failing for the hhvm version. which seems to be a problem with travis-ci (already on master branch) - my recommendation is to take hhvm out of .travis.yml because symfony 4 won't be supporting hhvm anymore.

rh389 commented:
Thanks - if you could merge master into this branch the tests ought to pass now (I've just bumped travis to trusty). I think we still need to test HHVM as long as we're supporting Symfony 2 and 3.

xabbuh commented on Twig/Extension/PhoneNumberHelperExtension.php:
```diff
@@ -0,0 +1,67 @@
+    {
+        return array(
+            new \Twig_SimpleFilter('phone_number_format', array($this->helper, 'format')),
+            new \Twig_SimpleFilter('phone_number_is_type', array($this->helper, 'isType')),
```
Usually, this would rather be a test in Twig templates. You would then use it like this:

```twig
{% if value is phone_number_type(type) %}
    {# ... #}
{% endif %}
```

toooni replied:
@xabbuh the problem in your example is that `value` always has to be the content of the constant. But the parameter for `phone_number_is_type` can also be a name of a constant in ``\libphonenumber\PhoneNumberFormat``.
I did this the same as the usage of `format`. You can take a look at https://github.com/misd-service-development/phone-number-bundle/pull/57/files#diff-549733e4205d824d550c9163b73ce246R110

xabbuh replied:
@toooni Well, it doesn't matter if you need to pass values to a filter, function or test. Can you show a Twig template using your filter that you have in mind that would not be writable with the test?

toooni replied:
@xabbuh oh. I didn't know that theres a `\Twig_SimpleTest`. I'll make the changes.

toooni replied:
@xabbuh is it ok when i name the test `phone_number_of_type`? And leave the deprecated function (I'll remove the filter).

xabbuh replied:
I have no preference for the name (but I also don't have anything to decide here ;)).

toooni replied:
But you're right. That should be enough ;) I made the changes - we'll see.

toooni commented:
@rh389 I switched to use a twig test instead of a twig filter as recommended by @xabbuh.

- [x] made a twig test instead of a twig filter
- [x] updated tests
- [x] updated README

rh389 commented:
That looks nice - more readable, cheers. And thanks for your help @xabbuh

rh389 commented:
Ok, merging - thanks again.

toooni commented:
thank you! have a nice day..
